docs/src/markdown.ts

```diff
@@ -0,0 +1,266 @@
+import fs from "fs/promises";
+import path from "path";
+import { unified } from "unified";
+import remarkParse from "remark-parse";
+import remarkRehype from "remark-rehype";
+import rehypeRaw from "rehype-raw";
+import { visit } from "unist-util-visit";
+import GitHubSlugger from "github-slugger";
+import matter from "gray-matter";
+
+export interface Document {
+  /** the Markdown file itself, without from-matter */
+  content: string;
+
+  /** the path to this markdown file */
+  path: string;
+
+  /** the headings found in this markdown file */
+  headings: string[];
+
+  frontMatter: {
+    title: string;
+    description: string;
+  };
+}
+
+export type ErrorType = "link" | "hash" | "source" | "related";
+
+export type LinkError = {
+  type: ErrorType;
+  href: string;
+  doc: Document;
+};
+
+export const DOCS_PATH = ".";
+export const EXCLUDED_HASHES = ["top"];
+
+const slugger = new GitHubSlugger();
+
+/** Collect the paths of all .mdx files in the passed directories */
+const getAllMdxFilePaths = async (
+  directoriesToScan: string[],
+  fileList: string[] = []
+): Promise<string[]> => {
+  for (const dir of directoriesToScan) {
+    const dirPath = path.join(".", dir);
+    const files = await fs.readdir(dirPath);
```

Comment:
Totally nonblocking and not worth a refactor, but was there a reason not to use the built in recursion?
```suggestion
    const files = await fs.readdir(dirPath, { recursive: true });
```

Reply:
this was overhauled in ef2d43be6bda7984cec9dc321e3e9d0342a1d715